File: MultilanguageSDK/Python/Samples/NYX650/AIGetFrameAndResultSync/AIGetFrameAndResultSync.py
from pickle import FALSE, TRUE
import sys
import threading
from queue import Queue
currentPath =  sys.path[0]
pos = currentPath.find('Samples')
libpath = currentPath[:pos]
print(libpath + "API")
sys.path.append(libpath) #absolutely path

from API.ScepterDS_api import *
import time
print("---AIGetFrameAndResultSync---")
camera = ScepterTofCam()

frameMutex = threading.Lock()
resultMutex = threading.Lock()
QueueMaxNum = 6
frameQueue = []
resultQueue = []
# frameQueue = Queue(QueueMaxNum) # dose not support front method
# resultQueue = Queue(QueueMaxNum)


class DealThread(threading.Thread):
    def __init__(self, type):
        super(DealThread, self).__init__()
        self.type = type
        self.running_ = True
    def run(self):
        if self.type == 1:
            #print("Frame Thread Running")
            while self.running_ :
                depthFrame = ScFrame()
                frameReady = ScFrameReady()
                ret, frameReady = camera.scGetFrameReady(c_uint16(1200))
                if  ret != 0:
                    print("[scGetFrameReady] fail, ScStatus({})".format(ret))
                    continue       
                if  frameReady.depth:
                    # must enable depthFrame before
                    ret, depthFrame = camera.scGetFrame(ScFrameType.SC_DEPTH_FRAME)
                    if  ret == 0:
                        frameMutex.acquire()
                        if  len(frameQueue) > QueueMaxNum:
                            frameQueue.pop(0)
                            frameQueue.append(depthFrame)
                        else:
                            frameQueue.append(depthFrame)
                        frameMutex.release()
                    else:
                        print("[scGetFrame] fail, ScStatus({})".format(ret))
        else:
            #print("AI Result Thread Running")
            while self.running_ :
                ret, result = camera.scAIModuleGetResult(waitTime = c_uint32(1200))
                if  ret != 0:
                    print("[scAIModuleGetResult] fail, ScStatus({})".format(ret))
                    continue
                resultMutex.acquire()
                if len(resultQueue) > QueueMaxNum:
                    resultQueue.pop(0)
                    resultQueue.append(result)
                else:
                    resultQueue.append(result)
                resultMutex.release()

def setParameters():
    # Set AI module continuous running.
    ret = camera.scAIModuleSetWorkMode(ScAIModuleMode.AI_CONTINUOUS_RUN_MODE)
    if ret == 0:
        print("[scAIModuleSetWorkMode] success, ScStatus({}). Set AI_CONTINUOUS_RUN_MODE.".format(ret))
    else:
        print("[scAIModuleSetWorkMode] fail, ScStatus({})".format(ret))
        return ret

	# Enable AI module.
    ret = camera.scAIModuleSetEnabled(c_bool(True))
    if ret == 0:
        print("[scAIModuleSetEnabled] success, ScStatus({}). Enable AI module.".format(ret))
    else:
        print("[scAIModuleSetEnabled] fail, ScStatus({})".format(ret))
        return ret
    return 0

camera_count = camera.scGetDeviceCount(3000)
print("[scGetDeviceCount] success, ScStatus(0), The device count is {}".format(camera_count))
if camera_count <= 0:
    print("[scGetDeviceCount] scans for 3000ms and then returns the device count is 0. Make sure the device is on the network before running the samples.")
    exit(1)

device_info=ScDeviceInfo()

if camera_count > 0:
    ret,device_infolist=camera.scGetDeviceInfoList(camera_count)
    if ret==0:
        device_info = device_infolist[0]
        print("[scGetDeviceInfoList] success, ScStatus({}). The first deviceInfo, <serialNumber>: {} , <ip>: {} , <status>: {}".format(ret, str(device_info.serialNumber.decode()), str(device_info.ip.decode()), str(device_info.status)))
        if  ScConnectStatus.SC_CONNECTABLE.value != device_info.status:
            print(" The first device [status]: {} does not support connection.".format(str(device_info.status)))
            exit(1)
    else:
        print("[scGetDeviceInfoList] fail, ScStatus({})".format(ret))
        exit(1) 
else: 
    print("there are no camera found")
    exit(1)

ret = camera.scOpenDeviceBySN(device_info.serialNumber)
if  ret == 0:
    print('[scOpenDeviceBySN] success ScStatus({}).'.format(str(ret)))
else:
    print('[scOpenDeviceBySN] fail ScStatus({}).'.format(str(ret)))
    exit(1)

ret = setParameters()
if  ret == 0:
    print('[setParameters] success ScStatus({}).'.format(str(ret)))
else:
    print('[setParameters] fail ScStatus({}).'.format(str(ret)))
    exit(1)

pbuffer = ctypes.c_void_p()
pbufferSize = ctypes.c_uint16(0)
ret = camera.scAIModuleGetParam(paramID = c_uint32(0), pBuffer = pbuffer, bufferSize = pbufferSize)
if ret == 0:
    str_buffer = ctypes.string_at(pbuffer, pbufferSize.value).decode('ascii')
    print("[scAIModuleGetParam] success ScStatus({}). The algo's version is: {}".format(str(ret), str(str_buffer)))
else:
    print('[scAIModuleGetParam] fail ScStatus({}).'.format(str(ret)))
    exit(1)

ret = camera.scAIModuleGetParam(paramID = c_uint32(1), pBuffer = pbuffer, bufferSize = pbufferSize)
if ret == 0:
    str_buffer = ctypes.string_at(pbuffer, pbufferSize.value).decode('ascii')
    print("[scAIModuleGetParam] success ScStatus({}). The algo's name is: {}".format(str(ret), str(str_buffer)))
else:
    print('[scAIModuleGetParam] fail ScStatus({}).'.format(str(ret)))
    exit(1)

ret, frameRate = camera.scGetFrameRate()
if  ret == 0:
    print('[scGetFrameRate] success ScStatus({}). The device frame rate is {}.'.format(str(ret), frameRate))
else:
    print('[scGetFrameRate] fail ScStatus({}).'.format(str(ret)))
    exit(1)

ret = camera.scStartStream()
if  ret == 0:
    print('[scStartStream] success ScStatus({}).'.format(str(ret)))
else:
    print('[scStartStream] fail ScStatus({}).'.format(str(ret)))
    exit(1) 


# sync
imageThread = DealThread(1)
imageThread.start()
resThread = DealThread(2)
resThread.start()

syncCount = 100
while syncCount > 0:
    frameMutex.acquire()
    resultMutex.acquire()
    while len(frameQueue) > 0 and len(resultQueue) > 0:
        frame = frameQueue[0]
        result = resultQueue[0]
        if frame.hardwaretimestamp == result.resultTimestamp:
            print("Sync frame index:", frame.frameIndex ,", result index: " , result.resultIndex , ", timestamp: " , result.resultTimestamp)
            frameQueue.pop(0)
            resultQueue.pop(0)
            syncCount = syncCount - 1
        elif frame.hardwaretimestamp > result.resultTimestamp:
            print("Sync result not match bigger, resultIndex: " , result.resultIndex , ", timestamp: " , result.resultTimestamp , ", frameIndex:" , frame.frameIndex , ", timestamp: " , frame.hardwaretimestamp)
            resultQueue.pop(0)
        else:
            print("Sync result not match less , resultIndex: " , result.resultIndex , ", timestamp: " , result.resultTimestamp , ", frameIndex:" , frame.frameIndex , ", timestamp: " , frame.hardwaretimestamp)
            frameQueue.pop(0)
    resultMutex.release()
    frameMutex.release()
    time.sleep(0.01)
imageThread.running_ = False
resThread.running_ = False
imageThread.join()
resThread.join()

ret = camera.scStopStream()
if  ret == 0:
    print('[scStopStream] success ScStatus({}).'.format(str(ret)))
else:
    print('[scStopStream] fail ScStatus({}).'.format(str(ret)))
    exit(1)

ret = camera.scCloseDevice()
if  ret == 0:
    print('[scCloseDevice] success ScStatus({}).'.format(str(ret)))
else:
    print('[scCloseDevice] fail ScStatus({}).'.format(str(ret)))
    exit(1)

exit(0)

File: BaseSDK/Ubuntu/Samples/Base/NYX650/AIGetFrameAndResultSync/AIGetFrameAndResultSync.cpp
#include <thread>
#include <iostream>
#include <iomanip>
#include <vector>
#include <mutex>
#include <condition_variable>
#include <queue>
#include <fstream>
#include <sys/timeb.h>
#include "Scepter_api.h"
#include "Scepter_Morph_api.h"

#define ResultCount 100
#define QueueMaxNum	6
using namespace std;

mutex frameMutex;
mutex resultMutex;
std::queue<ScFrame> frameQueue;
std::queue<ScAIResult> resultQueue;

class DealThread
{
public:
	void Deal(int type)
	{

		lock_guard<mutex> lk(mutex_);
		// type: 1. get frame;2. get result of AI mode.
		if (1 == type)
		{
			while (running_)
			{
				ScFrame depthFrame = { 0 };
				ScFrameReady frameReady = { 0 };
				ScStatus status = scGetFrameReady(device_, 1000, &frameReady);
				if (status != SC_OK)
				{
					cout << "[scGetFrameReady] fail, ScStatus(" << status << ")." << endl;
					continue;
				}

				// Get depth frame.
				if (1 == frameReady.depth)
				{
					status = scGetFrame(device_, SC_DEPTH_FRAME, &depthFrame);
					if (status == SC_OK && depthFrame.pFrameData != NULL)
					{
						lock_guard<mutex> lkFrame(frameMutex);
						if (frameQueue.size() >= QueueMaxNum)
						{
							frameQueue.pop();
							frameQueue.push(depthFrame);
						}
						else
						{
							frameQueue.push(depthFrame);
						}
					}
					else
					{
						cout << "[scGetFrame] fail, ScStatus(" << status << ")." << endl;
					}
				}
			};
		}
		else
		{
			while (running_)
			{
				ScAIResult aiResult = { 0 };
				ScStatus status = scAIModuleGetResult(device_, 1000, &aiResult);
				if (status != SC_OK)
				{
					cout << "[scAIModuleGetResult] fail, ScStatus(" << status << ")." << endl;
					continue;
				}
				{
					lock_guard<mutex> lkResult(resultMutex);
					if (resultQueue.size() >= QueueMaxNum)
					{
						resultQueue.pop();
						resultQueue.push(aiResult);
					}
					else
					{
						resultQueue.push(aiResult);
					}
				}
			};
		}

		isTestDone_ = true;
		condition_.notify_one();

		return;
	}

	bool WaitForTestDone()
	{
		unique_lock<mutex> lk(mutex_);
		condition_.wait(lk, [this] {return isTestDone_; });
		return isTestDone_;
	}

	void SetDeviceHandle(ScDeviceHandle device)
	{
		device_ = device;
	}

	void SetRunning(bool running)
	{
		running_ = running;
	}

private:
	ScDeviceHandle device_;
	bool isTestDone_ = false;
	bool running_ = true;
	mutex mutex_;
	condition_variable condition_;
};

ScStatus SetParams(ScDeviceHandle deviceHandle)
{
	ScStatus status = SC_OTHERS;
	status = scAIModuleSetPreviewFrameTypeEnabled(deviceHandle, SC_DEPTH_FRAME, true);
	if (status == ScStatus::SC_OK)
	{
		cout << "[scAIModuleSetPreviewFrameTypeEnabled] success, ScStatus(" << status << "). Preview SC_DEPTH_FRAME." << endl;
	}
	else
	{
		cout << "[scAIModuleSetPreviewFrameTypeEnabled] fail, ScStatus(" << status << ")." << endl;
		return status;
	}

	status = scAIModuleSetWorkMode(deviceHandle, AI_CONTINUOUS_RUN_MODE);
	if (status == ScStatus::SC_OK)
	{
		cout << "[scAIModuleSetWorkMode] success, ScStatus(" << status << "). Set AI_CONTINUOUS_RUN_MODE." << endl;
	}
	else
	{
		cout << "[scAIModuleSetWorkMode] fail, ScStatus(" << status << ")." << endl;
		return status;
	}

	status = scAIModuleSetEnabled(deviceHandle, true);
	if (status == ScStatus::SC_OK)
	{
		cout << "[scAIModuleSetEnabled] success, ScStatus(" << status << "). Enable AI module." << endl;
	}
	else
	{
		cout << "[scAIModuleSetEnabled] fail, ScStatus(" << status << ")." << endl;
		return status;
	}

	return SC_OK;
}

int main()
{
	cout << "---AIGetFrameAndResultSync---" << endl;

	uint32_t deviceCount;
	ScDeviceInfo* pDeviceListInfo = NULL;
	ScDeviceHandle deviceHandle = 0;
	ScStatus status = SC_OTHERS;

	status = scInitialize();
	if (status == ScStatus::SC_OK)
	{
		cout << "[scInitialize] success, ScStatus(" << status << ")." << endl;
	}
	else
	{
		cout << "[scInitialize] fail, ScStatus(" << status << ")." << endl;
		return 1;
	}

	status = scGetDeviceCount(&deviceCount, 3000);
	if (status == ScStatus::SC_OK)
	{
		cout << "[scGetDeviceCount] success, ScStatus(" << status << "). The device count is " << deviceCount << endl;
	}
	else
	{
		cout << "[scGetDeviceCount] fail, ScStatus(" << status << ")." << endl;
		return 1;
	}
	if (0 == deviceCount)
	{
		cout << "[scGetDeviceCount] scans for 3000ms and then returns the device count is 0. Make sure the device is on the network before running the samples." << endl;
		return 1;
	}

	pDeviceListInfo = new ScDeviceInfo[deviceCount];
	status = scGetDeviceInfoList(deviceCount, pDeviceListInfo);
	if (status == ScStatus::SC_OK)
	{
		cout << "[scGetDeviceInfoList] success, ScStatus(" << status << ").";
		if (SC_CONNECTABLE != pDeviceListInfo[0].status)
		{
			cout << " The first device [status]: " << pDeviceListInfo[0].status << " does not support connection." << endl;
			delete[] pDeviceListInfo;
			pDeviceListInfo = NULL;
			return 1;
		}
	}
	else
	{
		cout << "[scGetDeviceInfoList] fail, ScStatus(" << status << ")." << endl;
		delete[] pDeviceListInfo;
		pDeviceListInfo = NULL;
		return 1;
	}

	cout << " The first deviceInfo, <serialNumber>: " << pDeviceListInfo[0].serialNumber
		<< ", <ip>: " << pDeviceListInfo[0].ip << ", <status>: " << pDeviceListInfo[0].status << endl;

	status = scOpenDeviceBySN(pDeviceListInfo[0].serialNumber, &deviceHandle);
	if (status == ScStatus::SC_OK)
	{
		cout << "[scOpenDeviceBySN] success, ScStatus(" << status << ")." << endl;
		delete[] pDeviceListInfo;
		pDeviceListInfo = NULL;
	}
	else
	{
		cout << "[scOpenDeviceBySN] fail, ScStatus(" << status << ")." << endl;
		delete[] pDeviceListInfo;
		pDeviceListInfo = NULL;
		return 1;
	}
	
	void*    pBuffer = 0;
	uint16_t nBufferSize = 0;
	// Get algo version
	uint32_t paramID = 0;
	status = scAIModuleGetParam(deviceHandle, paramID, &pBuffer, &nBufferSize);
	if (status == ScStatus::SC_OK)
	{
		cout << "[scAIModuleGetParam] success, ScStatus(" << status << "). The algo's version is: " << (char*)pBuffer << endl;
	}
	else
	{
		cout << "[scAIModuleGetParam] fail, ScStatus(" << status << ")." << endl;
		return 1;
	}
	// Get algo name
	paramID = 1;
	status = scAIModuleGetParam(deviceHandle, paramID, &pBuffer, &nBufferSize);
	if (status == ScStatus::SC_OK)
	{
		cout << "[scAIModuleGetParam] success, ScStatus(" << status << "). The algo's name is: " << (char*)pBuffer << endl;
	}
	else
	{
		cout << "[scAIModuleGetParam] fail, ScStatus(" << status << ")." << endl;
		return 1;
	}

	//Set bSetParams to false. if the params has been initialized by "Parameter initialization file" in ScepterGUITool or by SDK API in "DeviceImportParamInitFile" sample.
	bool bSetParams = true;
	if (bSetParams)
	{
		status = SetParams(deviceHandle);
		if (status == ScStatus::SC_OK)
		{
			cout << "[SetParams] success, ScStatus(" << status << ")." << endl;
		}
		else
		{
			cout << "[SetParams] fail, ScStatus(" << status << ")." << endl;
			return 1;
		}
	}

	status = scStartStream(deviceHandle);
	if (status == ScStatus::SC_OK)
	{
		cout << "[scStartStream] success, ScStatus(" << status << ")." << endl;
	}
	else
	{
		cout << "[scStartStream] fail, ScStatus(" << status << ")." << endl;
		return 1;
	}

	DealThread* pDealThread = new DealThread[2];
	pDealThread[0].SetDeviceHandle(deviceHandle);
	pDealThread[1].SetDeviceHandle(deviceHandle);

	thread(&DealThread::Deal, &(pDealThread[0]), 1).detach();
	thread(&DealThread::Deal, &(pDealThread[1]), 2).detach();

	ScFrame tFrame;
	ScAIResult tAIResult;
	int syncCount = ResultCount;
	while (syncCount > 0)
	{
		{
			lock_guard<mutex> lkResult(resultMutex);
			lock_guard<mutex> lkFrame(frameMutex);
			while (!resultQueue.empty() && !frameQueue.empty())
			{
				tAIResult = resultQueue.front();
				tFrame = frameQueue.front();

				if (tFrame.deviceTimestamp == tAIResult.resultTimestamp)
				{
					cout << "Sync frame index:" << tFrame.frameIndex << ", result index: " << tAIResult.resultIndex << ", timestamp: " << tAIResult.resultTimestamp << endl;
					frameQueue.pop();
					resultQueue.pop();
					syncCount--;
					break;
				}
				else if (tFrame.deviceTimestamp > tAIResult.resultTimestamp)
				{
					cout << "Sync result not match bigger, resultIndex: " << tAIResult.resultIndex << ", timestamp: " << tAIResult.resultTimestamp<< ", frameIndex:" << tFrame.frameIndex << ", timestamp: " << tFrame.deviceTimestamp << endl;
					resultQueue.pop();
				}
				else if (tFrame.deviceTimestamp < tAIResult.resultTimestamp)
				{
					cout << "Sync frame not match less, frameIndex:" << tFrame.frameIndex << ", timestamp: " << tFrame.deviceTimestamp << ", resultIndex: " << tAIResult.resultIndex << ", timestamp : " << tAIResult.resultTimestamp<< endl;
					frameQueue.pop();
				}
			}
		}
		std::this_thread::sleep_for(std::chrono::milliseconds(10));
	}

	pDealThread[0].SetRunning(false);
	pDealThread[1].SetRunning(false);
	for (uint32_t i = 0; i < 2; i++)
	{
		pDealThread[i].WaitForTestDone();
	}
	delete[] pDealThread;

	status = scCloseDevice(&deviceHandle);
	if (status == ScStatus::SC_OK)
	{
		cout << "[scCloseDevice] success, ScStatus(" << status << ")." << endl;
	}
	else
	{
		cout << "[scCloseDevice] fail, ScStatus(" << status << ")." << endl;
		return 1;
	}

	status = scShutdown();
	if (status == ScStatus::SC_OK)
	{
		cout << "[scShutdown] success, ScStatus(" << status << ")." << endl;
	}
	else
	{
		cout << "[scShutdown] fail, ScStatus(" << status << ")." << endl;
		return 1;
	}
	cout << "---end---" << endl;

	return 0;
}
